File: java/mil/nga/giat/mage/sdk/gson/deserializer/EventDeserializer.java
package mil.nga.giat.mage.sdk.gson.deserializer;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;

import mil.nga.giat.mage.sdk.datastore.user.Event;

/**
 * JSON to {@link Event}
 *
 * @author wiedemanns
 *
 */
public class EventDeserializer implements JsonDeserializer<Event> {

    private static final String LOG_NAME = EventDeserializer.class.getName();

    /**
     * Convenience method for returning a Gson object with a registered GSon
     * TypeAdaptor i.e. custom deserializer.
     *
     * @return A Gson object that can be used to convert Json into a {@link Event}.
     */
    public static Gson getGsonBuilder() {
        GsonBuilder gsonBuilder = new GsonBuilder();
        gsonBuilder.registerTypeAdapter(Event.class, new EventDeserializer());
        return gsonBuilder.create();
    }

    @Override
    public Event deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        JsonObject event = json.getAsJsonObject();

        String remoteId = event.get("id").getAsString();

        String name = event.get("name").getAsString();

        String description = "";
        JsonElement descriptionElement = event.get("description");
        if (descriptionElement != null && !descriptionElement.isJsonNull()) {
            description = descriptionElement.getAsString();
        }

        String form = event.get("forms").toString();

        String acl = event.get("acl").toString();

        return new Event(remoteId, name, description, form, acl);
    }
}

File: mage/src/main/java/mil/nga/giat/mage/network/event/EventDeserializer.kt
package mil.nga.giat.mage.network.event

import com.google.gson.*
import mil.nga.giat.mage.network.gson.asJsonObjectOrNull
import mil.nga.giat.mage.network.gson.asStringOrNull
import mil.nga.giat.mage.database.model.event.Event
import mil.nga.giat.mage.database.model.event.Form
import java.lang.reflect.Type

class EventDeserializer : JsonDeserializer<Event> {

   @Throws(JsonParseException::class)
   override fun deserialize(
      json: JsonElement,
      typeOfT: Type,
      context: JsonDeserializationContext
   ): Event {
      val eventJson = json.asJsonObject

      val remoteId = eventJson.get("id").asString
      val name = eventJson.get("name").asString
      val description = eventJson.get("description")?.asJsonPrimitive?.asString
      val acl = eventJson["acl"].toString()
      val event = Event(
         remoteId,
         name,
         description,
         acl
      )
      eventJson.get("style")?.asJsonObjectOrNull()?.let { style ->
         event.style = style.toString()
      }

      val formsJson = eventJson.get("forms")
      if (formsJson?.isJsonArray == true) {
         event.forms = deserializeForms(formsJson.asJsonArray)
      }

      var minObservationForms: Int? = null
      val minObservationFormsElement = eventJson["minObservationForms"]
      if (minObservationFormsElement != null && minObservationFormsElement.isJsonPrimitive) {
         minObservationForms = minObservationFormsElement.asInt
      }
      event.minObservationForms = minObservationForms

      var maxObservationForms: Int? = null
      val maxObservationFormsElement = eventJson["maxObservationForms"]
      if (maxObservationFormsElement != null && maxObservationFormsElement.isJsonPrimitive) {
         maxObservationForms = maxObservationFormsElement.asInt
      }
      event.maxObservationForms = maxObservationForms

      return event
   }

   private fun deserializeForms(formsJson: JsonArray): List<Form> {
      val forms = mutableListOf<Form>()
      for (formJson in formsJson.filter { it.isJsonObject }) {
         forms.add(deserializeForm(formJson.asJsonObject))
      }

      return forms
   }

   private fun deserializeForm(formJson: JsonObject): Form {
      val form = Form()
      form.formId = formJson.get("id").asJsonPrimitive.asLong
      form.json = formJson.toString()

      formJson.get("primaryField")?.asStringOrNull()?.let {
         form.primaryMapField = it
      }

      formJson.get("variantField")?.asStringOrNull()?.let {
         form.secondaryMapField = it
      }

      formJson.get("primaryFeedField")?.asStringOrNull()?.let {
         form.primaryFeedField = it
      }

      formJson.get("secondaryFeedField")?.asStringOrNull()?.let {
         form.secondaryFeedField = it
      }

      formJson.get("style")?.asJsonObjectOrNull()?.let { style ->
         form.style = style.toString()
      }

      return form
   }

   companion object {
      /**
       * Convenience method for returning a Gson object with a registered GSon
       * TypeAdaptor i.e. custom deserializer.
       *
       * @return A Gson object that can be used to convert Json into a [Event].
       */
      @JvmStatic
      val gsonBuilder: Gson
         get() {
            val gsonBuilder = GsonBuilder()
            gsonBuilder.registerTypeAdapter(Event::class.java, EventDeserializer())
            return gsonBuilder.create()
         }
   }
}